### PR TITLE
add structured logger example app

### PR DESCRIPTION
**Pull request by briandowns**
Signed-off-by: Brian Downs <brian.downs@gmail.com>

### Well detailed description of the change :

Adds a simple JSON structured logger.

Resolves: #

#

### Type of change:

- [ ] Bug fix

- [ ] New feature

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping

- [ ] Tests have been updated to reflect the changes done within this PR (if applicable).

- [ ] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

### CODE REVIEW

File: examples/structured_logger.du [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        this.entry["timestamp"] = System.time();
+
+        if (fields) {
+            fields.keys().forEach(
+                def(entry) => this.entry[entry] = fields[entry]
+            );
+        }
+        this.entry["msg"] = msg;
+
+        var j = JSON.parse(this.entry.toString());
```
Comment (Jason2605) [inline examples/structured_logger.du:37]:
```suggestion
        const j = JSON.parse(this.entry.toString());
```

File: examples/structured_logger.du [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        if (this.output.len() > 1) {
+            with(this.output, "w") {
```
Comment (Jason2605) [inline examples/structured_logger.du:40]: We don't want to overwrite our logs!

```suggestion
            with(this.output, "a") {
```

File: examples/structured_logger.du [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    trace(msg, fields=nil) {
+        this.set_level("trace");
+        this.run(msg, fields);
+    }
+}
+
+// main
+{
+    var log = Log();
```
Comment (Jason2605) [inline examples/structured_logger.du:76]:
```suggestion
    const log = Log();
```

**Review by Jason2605 — CHANGES_REQUESTED** [review]
Massively nitpick comments really (sorry) but just a few changes!

I really like this example though, thanks!

**Comment by Jason2605** [comment]
Hm interestingly the CI/CD is exiting with 1. Not entirely sure why, it doesn't for me locally...

**Comment by briandowns** [comment]
> Hm interestingly the CI/CD is exiting with 1. Not entirely sure why, it doesn't for me locally...

Strange. it's doing the same for me as well. (macOS)

**Comment by Jason2605** [comment]
> > Hm interestingly the CI/CD is exiting with 1. Not entirely sure why, it doesn't for me locally...
> 
> Strange. it's doing the same for me as well. (macOS)

Hm. I'm on MacOS as well and it seems to exit as normal..

**Comment by briandowns** [comment]
@Jason2605 Looks like it was the exit call at the end of the example. Good to go.

**Comment by Jason2605** [comment]
> @Jason2605 Looks like it was the exit call at the end of the example. Good to go.

Thats made me a little uneasy haha. I tested locally and on an ubuntu VM and it didn't seem to exit with code 1. An exit(0) call also shouldn't have made it exit with code 1 🤯

**Comment by Jason2605** [comment]
Ahhh I know why it's giving exit code 1. The CI/CD is expecting the debug message "Total bytes lost: 0" but the `System.exit(0)` call is exiting out the VM early and not displaying this message.

**Pull request merged**